2211/2/tasks/TopBadWords.rb

```diff
@@ -0,0 +1,80 @@
+require 'russian_obscenity/base'
+require_relative '../MemberControl'
+require_relative '../WordsNumber'
+
+# 1st task
+class TopBadWords < MemberControl
+  attr_reader :members
+  attr_reader :number
+  def initialize
+    super
```

Comment:
Let's rewrite this `TopBadWords` class (and the rest of app) completely 😈.

Ok, right now we have many relatively small classes. So we don't have big methods, we rarely call private methods with arguments. Seems like that's what I wanted but it's not :).

The main problem in the code right now is manipulations with huge state and operating with primitives.

For example, we have many instance variables like `@members` and `@member`. And having variables like for all members and for a single member in one class kind of implies that we do something wrong. Because members is an array of single members.

The `TopBadWords` depends on many other classes and they depend on other classes. So if you decide to change something in this app you'll need to go and change many and many other lines of code.

Read more about dependencies between classes and why it is bad to have highly coupled classes in [POODR](http://www.poodr.com/).


Let me show you how I would design this class. 

1. The desired result of `TopBadWords` is to print a table of battlers (you call them members):
```rb
class TopBadWords
  def initialize(limit)
  end

  def print_members
  end
end
```

2. Let's expand `#print_members`:
```rb
class TopBadWords
  def print_members
    sorted_members.first(limit).each do |member|
      puts "#{member.name} | #{member.number_of_battles} | #{member.number_of_bad_words}"
    end
  end
end
```
Yes, `#sorted_members` is not implemented yet. I use non-existing methods. I pretend they exist and use them and implement them later.
Notice also that I pretend that `#sorted_members` returns an array of objects and those objects have methods like `#name` and `#number_of_battles`.

3. Let's implement fetching members.
```rb
class TopBadWords
  def sorted_members
    members.sort_by # ...
  end

  def members
    MembersLoader.new.members
  end
end
```

I pretend there is `MembersLoader` exist but it's not.

4. Let's implement this members loader:
```rb
class MembersLoader
  def members
    texts_by_name.map do |name, texts|
      Member.new(name, texts)
    end
  end

  def texts_by_name
    loaded_texts.group_by(&:name)
  end

  def loaded_texts
    Dir[...].each |filename|
      LoadedText.new(filename)
    end
  end
end
```

I wrote a lot. You don't need to go the exact way I described. I just tried to show how I think and how I design classes.